# <span style="color:blue"> Lending Club Case Study

### <span style="color:green"> First, let's import all necessary libraties

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings

%matplotlib inline
warnings.filterwarnings(action='ignore')

In [2]:
df = pd.read_csv("loan.csv")
df.head(10)

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   
3  1076863    1277178      10000        10000          10000.0   36 months   
4  1075358    1311748       3000         3000           3000.0   60 months   
5  1075269    1311441       5000         5000           5000.0   36 months   
6  1069639    1304742       7000         7000           7000.0   60 months   
7  1072053    1288686       3000         3000           3000.0   36 months   
8  1071795    1306957       5600         5600           5600.0   60 months   
9  1071570    1306721       5375         5375           5350.0   60 months   

  int_rate  installment grade sub_grade  ... num_tl_90g_dpd_24m  \
0   10.65%       162.87     B        B2  ...                NaN   
1   15.27%        59.83     C        C4  ...                NaN   
2   15.96%        84.33     C        C5  ...                NaN   
3   13.49%       339.31     C        C1  ...                NaN   
4   12.69%        67.79     B        B5  ...                NaN   
5    7.90%       156.46     A        A4  ...                NaN   
6   15.96%       170.08     C        C5  ...                NaN   
7   18.64%       109.43     E        E1  ...                NaN   
8   21.28%       152.39     F        F2  ...                NaN   
9   12.69%       121.45     B        B5  ...                NaN   

  num_tl_op_past_12m pct_tl_nvr_dlq  percent_bc_gt_75 pub_rec_bankruptcies  \
0                NaN            NaN               NaN                  0.0   
1                NaN            NaN               NaN                  0.0   
2                NaN            NaN               NaN                  0.0   
3                NaN            NaN               NaN                  0.0   
4                NaN            NaN               NaN                  0.0   
5                NaN            NaN               NaN                  0.0   
6                NaN            NaN               NaN                  0.0   
7                NaN            NaN               NaN                  0.0   
8                NaN            NaN               NaN                  0.0   
9                NaN            NaN               NaN                  0.0   

  tax_liens tot_hi_cred_lim total_bal_ex_mort total_bc_limit  \
0       0.0             NaN               NaN            NaN   
1       0.0             NaN               NaN            NaN   
2       0.0             NaN               NaN            NaN   
3       0.0             NaN               NaN            NaN   
4       0.0             NaN               NaN            NaN   
5       0.0             NaN               NaN            NaN   
6       0.0             NaN               NaN            NaN   
7       0.0             NaN               NaN            NaN   
8       0.0             NaN               NaN            NaN   
9       0.0             NaN               NaN            NaN   

  total_il_high_credit_limit  
0                        NaN  
1                        NaN  
2                        NaN  
3                        NaN  
4                        NaN  
5                        NaN  
6                        NaN  
7                        NaN  
8                        NaN  
9                        NaN  

[10 rows x 111 columns]

### <span style="color:green"> First thing is to have a brief look at the data and the null values etc.

### <span style="color:green"> We can see that, there are a lot columns with all 0 values and those columns are of no use to use in our analysis.

### <span style="color:green"> We can thus, single them out and delete

In [3]:
df.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39717 entries, 0 to 39716
Data columns (total 111 columns):
 #    Column                          Non-Null Count  Dtype  
---   ------                          --------------  -----  
 0    id                              39717 non-null  int64  
 1    member_id                       39717 non-null  int64  
 2    loan_amnt                       39717 non-null  int64  
 3    funded_amnt                     39717 non-null  int64  
 4    funded_amnt_inv                 39717 non-null  float64
 5    term                            39717 non-null  object 
 6    int_rate                        39717 non-null  object 
 7    installment                     39717 non-null  float64
 8    grade                           39717 non-null  object 
 9    sub_grade                       39717 non-null  object 
 10   emp_title                       37258 non-null  object 
 11   emp_length                      38642 non-null  object 
 12   home_ownership  

In [4]:
null_list=df.columns[df.isnull().all()].to_list()

In [5]:
null_list

['mths_since_last_major_derog',
 'annual_inc_joint',
 'dti_joint',
 'verification_status_joint',
 'tot_coll_amt',
 'tot_cur_bal',
 'open_acc_6m',
 'open_il_6m',
 'open_il_12m',
 'open_il_24m',
 'mths_since_rcnt_il',
 'total_bal_il',
 'il_util',
 'open_rv_12m',
 'open_rv_24m',
 'max_bal_bc',
 'all_util',
 'total_rev_hi_lim',
 'inq_fi',
 'total_cu_tl',
 'inq_last_12m',
 'acc_open_past_24mths',
 'avg_cur_bal',
 'bc_open_to_buy',
 'bc_util',
 'mo_sin_old_il_acct',
 'mo_sin_old_rev_tl_op',
 'mo_sin_rcnt_rev_tl_op',
 'mo_sin_rcnt_tl',
 'mort_acc',
 'mths_since_recent_bc',
 'mths_since_recent_bc_dlq',
 'mths_since_recent_inq',
 'mths_since_recent_revol_delinq',
 'num_accts_ever_120_pd',
 'num_actv_bc_tl',
 'num_actv_rev_tl',
 'num_bc_sats',
 'num_bc_tl',
 'num_il_tl',
 'num_op_rev_tl',
 'num_rev_accts',
 'num_rev_tl_bal_gt_0',
 'num_sats',
 'num_tl_120dpd_2m',
 'num_tl_30dpd',
 'num_tl_90g_dpd_24m',
 'num_tl_op_past_12m',
 'pct_tl_nvr_dlq',
 'percent_bc_gt_75',
 'tot_hi_cred_lim',
 'total_bal

In [6]:
df.drop(null_list, axis=1, inplace=True)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39717 entries, 0 to 39716
Data columns (total 57 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   id                          39717 non-null  int64  
 1   member_id                   39717 non-null  int64  
 2   loan_amnt                   39717 non-null  int64  
 3   funded_amnt                 39717 non-null  int64  
 4   funded_amnt_inv             39717 non-null  float64
 5   term                        39717 non-null  object 
 6   int_rate                    39717 non-null  object 
 7   installment                 39717 non-null  float64
 8   grade                       39717 non-null  object 
 9   sub_grade                   39717 non-null  object 
 10  emp_title                   37258 non-null  object 
 11  emp_length                  38642 non-null  object 
 12  home_ownership              39717 non-null  object 
 13  annual_inc                  397

### <span style="color:green"> Now, we are left with much lesser columns

### <span style="color:green"> Also, we can see that some of those columns are having some NULL values in them. We have to clean/treat them before moving ahead

In [8]:
df.columns[df.isnull().any()]

Index(['emp_title', 'emp_length', 'desc', 'title', 'mths_since_last_delinq',
       'mths_since_last_record', 'revol_util', 'last_pymnt_d', 'next_pymnt_d',
       'last_credit_pull_d', 'collections_12_mths_ex_med',
       'chargeoff_within_12_mths', 'pub_rec_bankruptcies', 'tax_liens'],
      dtype='object')

In [9]:
df['emp_title'].nunique()

28820

### <span style="color:green"> emp_title is the Job Title of the borrower and thus, can be one important factor in the analysis. We cannot delete the column.

### <span style="color:green"> We cannot use mode here as there are too many titles. We can rather delete the rows having the null values for emp_title

In [10]:
df[df['emp_title'].isnull()]

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
0      1077501    1296599       5000         5000           4975.0   
2      1077175    1313524       2400         2400           2400.0   
8      1071795    1306957       5600         5600           5600.0   
30     1032111    1261745       4375         4375           4375.0   
42     1069314    1304202       3000         3000           3000.0   
...        ...        ...        ...          ...              ...   
39700    98276      98268       5400         5400            200.0   
39704    93277      93254       3000         3000            950.0   
39706    92676      92671       5000         5000            150.0   
39714    90395      90390       5000         5000           1325.0   
39715    90376      89243       5000         5000            650.0   

             term int_rate  installment grade sub_grade  ... next_pymnt_d  \
0       36 months   10.65%       162.87     B        B2  ...          NaN   
2       36 months   15.96%        84.33     C        C5  ...          NaN   
8       60 months   21.28%       152.39     F        F2  ...          NaN   
30      36 months    7.51%       136.11     A        A3  ...          NaN   
42      36 months   18.25%       108.84     D        D5  ...          NaN   
...           ...      ...          ...   ...       ...  ...          ...   
39700   36 months    7.75%       168.60     A        A3  ...          NaN   
39704   36 months    8.70%        94.98     B        B1  ...          NaN   
39706   36 months    8.07%       156.84     A        A4  ...          NaN   
39714   36 months    8.07%       156.84     A        A4  ...          NaN   
39715   36 months    7.43%       155.38     A        A2  ...          NaN   

      last_credit_pull_d collections_12_mths_ex_med  policy_code  \
0                 May-16                        0.0            1   
2                 May-16                        0.0            1   
8                 Aug-12                        0.0            1   
30                Dec-14                        0.0            1   
42                May-16                        0.0            1   
...                  ...                        ...          ...   
39700             Dec-13                        NaN            1   
39704             Jul-10                        NaN            1   
39706             Jun-07                        NaN            1   
39714             Jun-07                        NaN            1   
39715             Jun-07                        NaN            1   

      application_type acc_now_delinq chargeoff_within_12_mths delinq_amnt  \
0           INDIVIDUAL              0                      0.0           0   
2           INDIVIDUAL              0                      0.0           0   
8           INDIVIDUAL              0                      0.0           0   
30          INDIVIDUAL              0                      0.0           0   
42          INDIVIDUAL              0                      0.0           0   
...                ...            ...                      ...         ...   
39700       INDIVIDUAL              0                      NaN           0   
39704       INDIVIDUAL              0                      NaN           0   
39706       INDIVIDUAL              0                      NaN           0   
39714       INDIVIDUAL              0                      NaN           0   
39715       INDIVIDUAL              0                      NaN           0   

      pub_rec_bankruptcies tax_liens  
0                      0.0       0.0  
2                      0.0       0.0  
8                      0.0       0.0  
30                     0.0       0.0  
42                     0.0       0.0  
...                    ...       ...  
39700                  NaN       NaN  
39704                  NaN       NaN  
39706                  NaN       NaN  
39714                  NaN       NaN  
39715                  NaN       NaN  

[2459 rows x 57 columns

In [11]:
df.drop(df[df['emp_title'].isnull()].index, inplace=True)

### <span style="color:green"> Let's check the data frame again

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 37258 entries, 1 to 39716
Data columns (total 57 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   id                          37258 non-null  int64  
 1   member_id                   37258 non-null  int64  
 2   loan_amnt                   37258 non-null  int64  
 3   funded_amnt                 37258 non-null  int64  
 4   funded_amnt_inv             37258 non-null  float64
 5   term                        37258 non-null  object 
 6   int_rate                    37258 non-null  object 
 7   installment                 37258 non-null  float64
 8   grade                       37258 non-null  object 
 9   sub_grade                   37258 non-null  object 
 10  emp_title                   37258 non-null  object 
 11  emp_length                  37202 non-null  object 
 12  home_ownership              37258 non-null  object 
 13  annual_inc                  37258 no

In [13]:
df.columns[df.isnull().any()]

Index(['emp_length', 'desc', 'title', 'mths_since_last_delinq',
       'mths_since_last_record', 'revol_util', 'last_pymnt_d', 'next_pymnt_d',
       'collections_12_mths_ex_med', 'chargeoff_within_12_mths',
       'pub_rec_bankruptcies', 'tax_liens'],
      dtype='object')

### <span style="color:green"> Let's deal with emp_length next (Employment length in years. Possible values are between 0 and 10 where 0 means less than one year and 10 means ten or more years)

In [14]:
df['emp_length'].unique()

array(['< 1 year', '10+ years', '1 year', '3 years', '8 years', '9 years',
       '5 years', '4 years', '6 years', '2 years', '7 years', nan],
      dtype=object)

In [15]:
df['emp_length'] = df['emp_length'].fillna(df['emp_length'].mode()[0])

In [16]:
df['emp_length'].unique()

array(['< 1 year', '10+ years', '1 year', '3 years', '8 years', '9 years',
       '5 years', '4 years', '6 years', '2 years', '7 years'],
      dtype=object)

In [17]:
df.columns[df.isnull().any()]

Index(['desc', 'title', 'mths_since_last_delinq', 'mths_since_last_record',
       'revol_util', 'last_pymnt_d', 'next_pymnt_d',
       'collections_12_mths_ex_med', 'chargeoff_within_12_mths',
       'pub_rec_bankruptcies', 'tax_liens'],
      dtype='object')

### <span style="color:green"> Let's deal with next feature desc (Loan description provided by the borrower)

In [18]:
df['desc'].head(10)

1       Borrower added on 12/22/11 > I plan to use t...
3       Borrower added on 12/21/11 > to pay for prop...
4       Borrower added on 12/21/11 > I plan on combi...
5                                                   NaN
6       Borrower added on 12/18/11 > I am planning o...
7       Borrower added on 12/16/11 > Downpayment for...
9       Borrower added on 12/16/11 > I'm trying to b...
10      Borrower added on 12/15/11 > I had recived a...
11                                                  NaN
12      Borrower added on 12/15/11 > Plan to pay off...
Name: desc, dtype: object

In [19]:
df['desc'].info()

<class 'pandas.core.series.Series'>
Index: 37258 entries, 1 to 39716
Series name: desc
Non-Null Count  Dtype 
--------------  ----- 
25171 non-null  object
dtypes: object(1)
memory usage: 582.2+ KB


### <span style="color:green"> This doesn't look like to add value to our analysis and also, there are so many missing values. It's better to drop the column

In [20]:
df = df.drop('desc',axis=1)

In [21]:
df.columns[df.isnull().any()]

Index(['title', 'mths_since_last_delinq', 'mths_since_last_record',
       'revol_util', 'last_pymnt_d', 'next_pymnt_d',
       'collections_12_mths_ex_med', 'chargeoff_within_12_mths',
       'pub_rec_bankruptcies', 'tax_liens'],
      dtype='object')

### <span style="color:green"> Let's deal with next feature title (The loan title provided by the borrower)

In [22]:
df['title'].head(10)

1                                      bike
3                                  personel
4                                  Personal
5     My wedding loan I promise to pay back
6                                      Loan
7                           Car Downpayment
9               Building my credit history.
10               High intrest Consolidation
11                            Consolidation
12                                  freedom
Name: title, dtype: object

In [23]:
df['title'].info()

<class 'pandas.core.series.Series'>
Index: 37258 entries, 1 to 39716
Series name: title
Non-Null Count  Dtype 
--------------  ----- 
37249 non-null  object
dtypes: object(1)
memory usage: 582.2+ KB


In [24]:
df['title'].nunique()

18417

### <span style="color:green"> Like the previous column, this also doesn't look like to add any value to our analysis. It's better to drop the column

In [25]:
df = df.drop('title', axis=1)

In [26]:
df.columns[df.isnull().any()]

Index(['mths_since_last_delinq', 'mths_since_last_record', 'revol_util',
       'last_pymnt_d', 'next_pymnt_d', 'collections_12_mths_ex_med',
       'chargeoff_within_12_mths', 'pub_rec_bankruptcies', 'tax_liens'],
      dtype='object')

### <span style="color:green"> Next variable/feature: mths_since_last_delinq (The number of months since the borrower's last delinquency)

In [27]:
df['mths_since_last_delinq'].info()

<class 'pandas.core.series.Series'>
Index: 37258 entries, 1 to 39716
Series name: mths_since_last_delinq
Non-Null Count  Dtype  
--------------  -----  
13197 non-null  float64
dtypes: float64(1)
memory usage: 582.2 KB


### <span style="color:green"> There are too many missing values in this feature. Though seems to be an important one, we cannot go ahead with this column due to so many missing data.

### <span style="color:green"> Some more such columns are mths_since_last_record (The number of months since the last public record) and next_pymnt_d (Next scheduled payment date)

### <span style="color:green"> The only logical thing to do is to delete these features.

In [28]:
df = df.drop(['mths_since_last_delinq', 'mths_since_last_record', 'next_pymnt_d'], axis=1)

In [29]:
df.columns[df.isnull().any()]

Index(['revol_util', 'last_pymnt_d', 'collections_12_mths_ex_med',
       'chargeoff_within_12_mths', 'pub_rec_bankruptcies', 'tax_liens'],
      dtype='object')

### <span style="color:green"> Next variable to deal with is revol_util (Revolving line utilization rate, or the amount of credit the borrower is using relative to all available revolving credit)

In [30]:
df['revol_util'].info()

<class 'pandas.core.series.Series'>
Index: 37258 entries, 1 to 39716
Series name: revol_util
Non-Null Count  Dtype 
--------------  ----- 
37212 non-null  object
dtypes: object(1)
memory usage: 582.2+ KB


In [31]:
df['revol_util'].nunique()

1082

### <span style="color:green"> There are a very less amount of missing data and we can delete the rows without losing much of information.

In [32]:
df.drop(df[df['revol_util'].isnull()].index, inplace=True)

In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 37212 entries, 1 to 39716
Data columns (total 52 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   id                          37212 non-null  int64  
 1   member_id                   37212 non-null  int64  
 2   loan_amnt                   37212 non-null  int64  
 3   funded_amnt                 37212 non-null  int64  
 4   funded_amnt_inv             37212 non-null  float64
 5   term                        37212 non-null  object 
 6   int_rate                    37212 non-null  object 
 7   installment                 37212 non-null  float64
 8   grade                       37212 non-null  object 
 9   sub_grade                   37212 non-null  object 
 10  emp_title                   37212 non-null  object 
 11  emp_length                  37212 non-null  object 
 12  home_ownership              37212 non-null  object 
 13  annual_inc                  37212 no

In [34]:
df.columns[df.isnull().any()]

Index(['last_pymnt_d', 'collections_12_mths_ex_med',
       'chargeoff_within_12_mths', 'pub_rec_bankruptcies', 'tax_liens'],
      dtype='object')

### <span style="color:green"> Next column is last_pymnt_d (Last month payment was received)

In [35]:
df['last_pymnt_d'].info()

<class 'pandas.core.series.Series'>
Index: 37212 entries, 1 to 39716
Series name: last_pymnt_d
Non-Null Count  Dtype 
--------------  ----- 
37152 non-null  object
dtypes: object(1)
memory usage: 581.4+ KB


In [36]:
df['last_pymnt_d'].unique()

array(['Apr-13', 'Jan-15', 'May-16', 'Nov-12', 'Jun-13', 'Sep-13',
       'Jul-12', 'Oct-13', 'May-13', 'Feb-15', 'Aug-15', 'Oct-12',
       'Sep-12', nan, 'Dec-12', 'Dec-14', 'Aug-13', 'Nov-13', 'Jan-14',
       'Apr-14', 'Aug-14', 'Oct-14', 'Aug-12', 'Jul-14', 'Jul-13',
       'Jan-16', 'Apr-15', 'Feb-14', 'Sep-14', 'Jun-12', 'Feb-13',
       'Mar-13', 'May-14', 'Mar-15', 'Jan-13', 'Dec-13', 'Jun-14',
       'Feb-12', 'Mar-14', 'Apr-12', 'Sep-15', 'Nov-15', 'Mar-16',
       'Feb-16', 'Jan-12', 'Oct-15', 'Nov-14', 'Mar-12', 'May-12',
       'Apr-16', 'Dec-15', 'Jun-15', 'May-15', 'Jul-15', 'Dec-11',
       'Nov-11', 'Oct-11', 'Sep-11', 'Aug-11', 'Jul-11', 'Jun-11',
       'May-11', 'Apr-11', 'Mar-11', 'Feb-11', 'Jan-11', 'Dec-10',
       'Nov-10', 'Oct-10', 'Sep-10', 'Aug-10', 'Jul-10', 'Jun-10',
       'May-10', 'Apr-10', 'Mar-10', 'Feb-10', 'Jan-10', 'Dec-09',
       'Nov-09', 'Oct-09', 'Sep-09', 'Aug-09', 'Jul-09', 'Jun-09',
       'May-09', 'Apr-09', 'Mar-09', 'Feb-09', 'Jan-09', 

### <span style="color:green"> Like the last column, there are a very less amount of missing data and we can delete the rows without losing much of information.

In [37]:
df.drop(df[df['last_pymnt_d'].isnull()].index, inplace=True)

In [38]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 37152 entries, 1 to 39716
Data columns (total 52 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   id                          37152 non-null  int64  
 1   member_id                   37152 non-null  int64  
 2   loan_amnt                   37152 non-null  int64  
 3   funded_amnt                 37152 non-null  int64  
 4   funded_amnt_inv             37152 non-null  float64
 5   term                        37152 non-null  object 
 6   int_rate                    37152 non-null  object 
 7   installment                 37152 non-null  float64
 8   grade                       37152 non-null  object 
 9   sub_grade                   37152 non-null  object 
 10  emp_title                   37152 non-null  object 
 11  emp_length                  37152 non-null  object 
 12  home_ownership              37152 non-null  object 
 13  annual_inc                  37152 no

In [39]:
df.columns[df.isnull().any()]

Index(['collections_12_mths_ex_med', 'chargeoff_within_12_mths',
       'pub_rec_bankruptcies', 'tax_liens'],
      dtype='object')

### <span style="color:green"> Next Feature: collections_12_mths_ex_med (Number of collections in 12 months excluding medical collections)

In [40]:
df['collections_12_mths_ex_med'].info()

<class 'pandas.core.series.Series'>
Index: 37152 entries, 1 to 39716
Series name: collections_12_mths_ex_med
Non-Null Count  Dtype  
--------------  -----  
37105 non-null  float64
dtypes: float64(1)
memory usage: 580.5 KB


In [41]:
df['collections_12_mths_ex_med'].unique()

array([ 0., nan])

In [42]:
df['collections_12_mths_ex_med'].head(10)

1     0.0
3     0.0
4     0.0
5     0.0
6     0.0
7     0.0
9     0.0
10    0.0
11    0.0
12    0.0
Name: collections_12_mths_ex_med, dtype: float64

### <span style="color:green"> This Column contains all 0's and few Null values. We can easily delete this column

In [43]:
df = df.drop('collections_12_mths_ex_med', axis=1)

In [44]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 37152 entries, 1 to 39716
Data columns (total 51 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   id                        37152 non-null  int64  
 1   member_id                 37152 non-null  int64  
 2   loan_amnt                 37152 non-null  int64  
 3   funded_amnt               37152 non-null  int64  
 4   funded_amnt_inv           37152 non-null  float64
 5   term                      37152 non-null  object 
 6   int_rate                  37152 non-null  object 
 7   installment               37152 non-null  float64
 8   grade                     37152 non-null  object 
 9   sub_grade                 37152 non-null  object 
 10  emp_title                 37152 non-null  object 
 11  emp_length                37152 non-null  object 
 12  home_ownership            37152 non-null  object 
 13  annual_inc                37152 non-null  float64
 14  verificatio

In [45]:
df.columns[df.isnull().any()]

Index(['chargeoff_within_12_mths', 'pub_rec_bankruptcies', 'tax_liens'], dtype='object')

### <span style="color:green"> Next, let's examine column chargeoff_within_12_mths (Number of charge-offs within 12 months)

In [46]:
df['chargeoff_within_12_mths'].info()

<class 'pandas.core.series.Series'>
Index: 37152 entries, 1 to 39716
Series name: chargeoff_within_12_mths
Non-Null Count  Dtype  
--------------  -----  
37105 non-null  float64
dtypes: float64(1)
memory usage: 580.5 KB


In [47]:
df['chargeoff_within_12_mths'].unique()

array([ 0., nan])

### <span style="color:green"> Like the previous columns, this column also is having all 0's in it apart from few Null values. Column can thus be deleted.

In [48]:
df = df.drop('chargeoff_within_12_mths', axis=1)

In [49]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 37152 entries, 1 to 39716
Data columns (total 50 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       37152 non-null  int64  
 1   member_id                37152 non-null  int64  
 2   loan_amnt                37152 non-null  int64  
 3   funded_amnt              37152 non-null  int64  
 4   funded_amnt_inv          37152 non-null  float64
 5   term                     37152 non-null  object 
 6   int_rate                 37152 non-null  object 
 7   installment              37152 non-null  float64
 8   grade                    37152 non-null  object 
 9   sub_grade                37152 non-null  object 
 10  emp_title                37152 non-null  object 
 11  emp_length               37152 non-null  object 
 12  home_ownership           37152 non-null  object 
 13  annual_inc               37152 non-null  float64
 14  verification_status      37

In [50]:
df.columns[df.isnull().any()]

Index(['pub_rec_bankruptcies', 'tax_liens'], dtype='object')

### <span style="color:green"> Let's examine the column pub_rec_bankruptcies (Number of public record bankruptcies) next

In [51]:
df['pub_rec_bankruptcies'].info()

<class 'pandas.core.series.Series'>
Index: 37152 entries, 1 to 39716
Series name: pub_rec_bankruptcies
Non-Null Count  Dtype  
--------------  -----  
36498 non-null  float64
dtypes: float64(1)
memory usage: 580.5 KB


In [52]:
df['pub_rec_bankruptcies'].unique()

array([ 0.,  1.,  2., nan])

In [53]:
df['pub_rec_bankruptcies'].nunique()

3

### <span style="color:green"> For this column, we can replace the missing values with the Mode

In [54]:
df['pub_rec_bankruptcies'].fillna(df['pub_rec_bankruptcies'].mode()[0],inplace=True)

In [55]:
df.columns[df.isnull().any()]

Index(['tax_liens'], dtype='object')

### <span style="color:green"> Only remaining feature with Null values now is tax_liens (Number of tax liens)

In [56]:
df['tax_liens'].info()

<class 'pandas.core.series.Series'>
Index: 37152 entries, 1 to 39716
Series name: tax_liens
Non-Null Count  Dtype  
--------------  -----  
37120 non-null  float64
dtypes: float64(1)
memory usage: 580.5 KB


In [57]:
df['tax_liens'].unique()

array([ 0., nan])

### <span style="color:green"> This column contains only 0's and few Null values. Can be deleted.

In [58]:
df = df.drop('tax_liens', axis=1)

In [59]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 37152 entries, 1 to 39716
Data columns (total 49 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       37152 non-null  int64  
 1   member_id                37152 non-null  int64  
 2   loan_amnt                37152 non-null  int64  
 3   funded_amnt              37152 non-null  int64  
 4   funded_amnt_inv          37152 non-null  float64
 5   term                     37152 non-null  object 
 6   int_rate                 37152 non-null  object 
 7   installment              37152 non-null  float64
 8   grade                    37152 non-null  object 
 9   sub_grade                37152 non-null  object 
 10  emp_title                37152 non-null  object 
 11  emp_length               37152 non-null  object 
 12  home_ownership           37152 non-null  object 
 13  annual_inc               37152 non-null  float64
 14  verification_status      37

### <span style="color:blue"> Now, we are left with 49 columns and 37152 rows

### <span style="color:green"> Let's next check if we have any rows having all 0's. The rows can be deleted as well.

In [60]:
df[(df==0).all(axis=1)]

Empty DataFrame
Columns: [id, member_id, loan_amnt, funded_amnt, funded_amnt_inv, term, int_rate, installment, grade, sub_grade, emp_title, emp_length, home_ownership, annual_inc, verification_status, issue_d, loan_status, pymnt_plan, url, purpose, zip_code, addr_state, dti, delinq_2yrs, earliest_cr_line, inq_last_6mths, open_acc, pub_rec, revol_bal, revol_util, total_acc, initial_list_status, out_prncp, out_prncp_inv, total_pymnt, total_pymnt_inv, total_rec_prncp, total_rec_int, total_rec_late_fee, recoveries, collection_recovery_fee, last_pymnt_d, last_pymnt_amnt, last_credit_pull_d, policy_code, application_type, acc_now_delinq, delinq_amnt, pub_rec_bankruptcies]
Index: []

[0 rows x 49 columns]

### <span style="color:green"> Well, there are no such rows.

### <span style="color:green"> Let's now check if the features need any data type modifications

In [61]:
df['term'].unique()

array([' 60 months', ' 36 months'], dtype=object)

### <span style="color:green">  We will remove 'months' and make it interger

In [62]:
def remove_char(x):
    return int(x.strip().split(' ')[0])

In [63]:
df['term'] = df['term'].apply(remove_char)

In [64]:
df['term'].unique()

array([60, 36], dtype=int64)

In [65]:
df['term'].info()

<class 'pandas.core.series.Series'>
Index: 37152 entries, 1 to 39716
Series name: term
Non-Null Count  Dtype
--------------  -----
37152 non-null  int64
dtypes: int64(1)
memory usage: 580.5 KB


In [66]:
df['int_rate'].unique()

array(['15.27%', '13.49%', '12.69%', '7.90%', '15.96%', '18.64%',
       '14.65%', '9.91%', '10.65%', '16.29%', '6.03%', '11.71%', '12.42%',
       '16.77%', '8.90%', '7.51%', '6.62%', '19.91%', '17.27%', '14.27%',
       '21.28%', '17.58%', '21.67%', '19.42%', '22.06%', '20.89%',
       '18.25%', '20.30%', '23.91%', '19.03%', '23.52%', '23.13%',
       '22.74%', '22.35%', '24.11%', '6.00%', '22.11%', '7.49%', '11.99%',
       '5.99%', '10.99%', '9.99%', '18.79%', '11.49%', '15.99%', '16.49%',
       '6.99%', '12.99%', '15.23%', '14.79%', '8.49%', '5.42%', '17.49%',
       '10.59%', '15.62%', '21.36%', '19.29%', '13.99%', '18.39%',
       '16.89%', '17.99%', '20.62%', '20.99%', '22.85%', '19.69%',
       '20.25%', '23.22%', '21.74%', '22.48%', '23.59%', '12.62%',
       '18.07%', '11.63%', '7.91%', '7.42%', '11.14%', '20.20%', '12.12%',
       '19.39%', '16.11%', '17.54%', '16.59%', '17.19%', '12.87%',
       '20.69%', '9.67%', '21.82%', '19.79%', '18.49%', '13.84%',
       '22.94%', '

### <span style="color:green"> We can remove % and make it float

In [67]:
df['int_rate'] = df['int_rate'].apply(lambda x: float(x.rstrip('%')))

In [68]:
df['int_rate'].unique()

array([15.27, 13.49, 12.69,  7.9 , 15.96, 18.64, 14.65,  9.91, 10.65,
       16.29,  6.03, 11.71, 12.42, 16.77,  8.9 ,  7.51,  6.62, 19.91,
       17.27, 14.27, 21.28, 17.58, 21.67, 19.42, 22.06, 20.89, 18.25,
       20.3 , 23.91, 19.03, 23.52, 23.13, 22.74, 22.35, 24.11,  6.  ,
       22.11,  7.49, 11.99,  5.99, 10.99,  9.99, 18.79, 11.49, 15.99,
       16.49,  6.99, 12.99, 15.23, 14.79,  8.49,  5.42, 17.49, 10.59,
       15.62, 21.36, 19.29, 13.99, 18.39, 16.89, 17.99, 20.62, 20.99,
       22.85, 19.69, 20.25, 23.22, 21.74, 22.48, 23.59, 12.62, 18.07,
       11.63,  7.91,  7.42, 11.14, 20.2 , 12.12, 19.39, 16.11, 17.54,
       16.59, 17.19, 12.87, 20.69,  9.67, 21.82, 19.79, 18.49, 13.84,
       22.94, 24.4 , 21.48, 14.82, 17.88, 20.11, 16.02, 17.51, 13.06,
       14.91, 15.28, 17.14, 13.43, 11.11, 10.37, 14.17, 16.4 , 15.65,
        7.66, 10.  , 10.74,  7.29,  5.79,  6.92,  9.63, 14.54, 12.68,
       18.62, 19.36, 13.8 , 18.99, 21.59, 20.85, 21.22, 19.74, 20.48,
        6.91, 12.23,

In [69]:
df['int_rate'].info()

<class 'pandas.core.series.Series'>
Index: 37152 entries, 1 to 39716
Series name: int_rate
Non-Null Count  Dtype  
--------------  -----  
37152 non-null  float64
dtypes: float64(1)
memory usage: 580.5 KB


### <span style="color:green"> Let's check some other variables one after the other

In [70]:
df['grade'].info()

<class 'pandas.core.series.Series'>
Index: 37152 entries, 1 to 39716
Series name: grade
Non-Null Count  Dtype 
--------------  ----- 
37152 non-null  object
dtypes: object(1)
memory usage: 580.5+ KB


In [71]:
df['grade'].unique()

array(['C', 'B', 'A', 'E', 'D', 'F', 'G'], dtype=object)

In [72]:
df['sub_grade'].unique()

array(['C4', 'C1', 'B5', 'A4', 'C5', 'E1', 'C3', 'B1', 'B2', 'D1', 'A1',
       'B3', 'B4', 'D2', 'A5', 'A3', 'A2', 'E4', 'D3', 'C2', 'F2', 'D4',
       'F3', 'E3', 'F4', 'F1', 'D5', 'E5', 'G4', 'E2', 'G3', 'G2', 'G1',
       'F5', 'G5'], dtype=object)

In [73]:
df['emp_title'].unique()

array(['Ryder', 'AIR RESOURCES BOARD', 'University Medical Group', ...,
       'FiSite Research', 'Squarewave Solutions, Ltd.',
       'Evergreen Center'], dtype=object)

In [74]:
df['emp_title'].nunique()

28734

In [75]:
df['emp_length'].unique()

array(['< 1 year', '10+ years', '1 year', '3 years', '8 years', '9 years',
       '5 years', '4 years', '6 years', '2 years', '7 years'],
      dtype=object)

In [76]:
df['home_ownership'].unique()

array(['RENT', 'OWN', 'MORTGAGE', 'OTHER', 'NONE'], dtype=object)

In [77]:
df['verification_status'].unique()

array(['Source Verified', 'Not Verified', 'Verified'], dtype=object)

In [78]:
df['issue_d'].unique()

array(['Dec-11', 'Nov-11', 'Oct-11', 'Sep-11', 'Aug-11', 'Jul-11',
       'Jun-11', 'May-11', 'Apr-11', 'Mar-11', 'Feb-11', 'Jan-11',
       'Dec-10', 'Nov-10', 'Oct-10', 'Sep-10', 'Aug-10', 'Jul-10',
       'Jun-10', 'May-10', 'Apr-10', 'Mar-10', 'Feb-10', 'Jan-10',
       'Dec-09', 'Nov-09', 'Oct-09', 'Sep-09', 'Aug-09', 'Jul-09',
       'Jun-09', 'May-09', 'Apr-09', 'Mar-09', 'Feb-09', 'Jan-09',
       'Dec-08', 'Nov-08', 'Oct-08', 'Sep-08', 'Aug-08', 'Jul-08',
       'Jun-08', 'May-08', 'Apr-08', 'Mar-08', 'Feb-08', 'Jan-08',
       'Dec-07', 'Nov-07', 'Oct-07', 'Sep-07', 'Aug-07', 'Jul-07',
       'Jun-07'], dtype=object)

### <span style="color:green"> We can do 2 things for this column. Change the column name to "issued" and separate month/year

In [79]:
df.rename(columns={'issue_d':'issued'}, inplace=True)

In [80]:
df['issued_month'] = df['issued'].apply(lambda x: x.split('-')[0])
df['issued_year'] = df['issued'].apply(lambda x: x.split('-')[1])
df['issued_year'] = df['issued_year'].apply(lambda x: '20'+ x)

In [81]:
df['issued_month'].head()

1    Dec
3    Dec
4    Dec
5    Dec
6    Dec
Name: issued_month, dtype: object

In [82]:
df['issued_year'].head()

1    2011
3    2011
4    2011
5    2011
6    2011
Name: issued_year, dtype: object

In [83]:
df['loan_status'].unique()

array(['Charged Off', 'Fully Paid', 'Current'], dtype=object)

In [84]:
df['pymnt_plan'].unique()

array(['n'], dtype=object)

### <span style="color:red"> The column 'pymnt_plan' has a singular value. It's of no use for our analysis. We can drop it.

In [85]:
df = df.drop('pymnt_plan', axis=1)

In [86]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 37152 entries, 1 to 39716
Data columns (total 50 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       37152 non-null  int64  
 1   member_id                37152 non-null  int64  
 2   loan_amnt                37152 non-null  int64  
 3   funded_amnt              37152 non-null  int64  
 4   funded_amnt_inv          37152 non-null  float64
 5   term                     37152 non-null  int64  
 6   int_rate                 37152 non-null  float64
 7   installment              37152 non-null  float64
 8   grade                    37152 non-null  object 
 9   sub_grade                37152 non-null  object 
 10  emp_title                37152 non-null  object 
 11  emp_length               37152 non-null  object 
 12  home_ownership           37152 non-null  object 
 13  annual_inc               37152 non-null  float64
 14  verification_status      37

In [87]:
df['url'].unique()

array(['https://lendingclub.com/browse/loanDetail.action?loan_id=1077430',
       'https://lendingclub.com/browse/loanDetail.action?loan_id=1076863',
       'https://lendingclub.com/browse/loanDetail.action?loan_id=1075358',
       ...,
       'https://lendingclub.com/browse/loanDetail.action?loan_id=92187',
       'https://lendingclub.com/browse/loanDetail.action?loan_id=90665',
       'https://lendingclub.com/browse/loanDetail.action?loan_id=87023'],
      dtype=object)

In [88]:
df['purpose'].unique()

array(['car', 'other', 'wedding', 'debt_consolidation', 'credit_card',
       'home_improvement', 'major_purchase', 'medical', 'moving',
       'small_business', 'vacation', 'house', 'renewable_energy',
       'educational'], dtype=object)

In [89]:
df['zip_code'].head(10)

1     309xx
3     917xx
4     972xx
5     852xx
6     280xx
7     900xx
9     774xx
10    853xx
11    913xx
12    245xx
Name: zip_code, dtype: object

In [90]:
df['addr_state'].unique()

array(['GA', 'CA', 'OR', 'AZ', 'NC', 'TX', 'VA', 'IL', 'MO', 'CT', 'UT',
       'FL', 'PA', 'MN', 'NJ', 'KY', 'OH', 'SC', 'RI', 'LA', 'MA', 'NY',
       'WA', 'WI', 'AL', 'CO', 'KS', 'NV', 'AK', 'MD', 'WV', 'VT', 'MI',
       'DC', 'NH', 'AR', 'NM', 'MT', 'HI', 'WY', 'OK', 'SD', 'DE', 'MS',
       'TN', 'IA', 'NE', 'ID', 'IN', 'ME'], dtype=object)

In [91]:
df['earliest_cr_line'].head(10)

1     Apr-99
3     Feb-96
4     Jan-96
5     Nov-04
6     Jul-05
7     Jan-07
9     Sep-04
10    Jan-98
11    Oct-89
12    Apr-04
Name: earliest_cr_line, dtype: object

### <span style="color:green"> We can separate month/year for this column as well. Also, we will rewrite the year for readability.

In [132]:
df['earliest_cr_line_month'] = df['earliest_cr_line'].apply(lambda x: x.split('-')[0])
df['earliest_cr_line_year'] = df['earliest_cr_line'].apply(lambda x: x.split('-')[1])

In [133]:
df['earliest_cr_line_month'].head()

1    Apr
3    Feb
4    Jan
5    Nov
6    Jul
Name: earliest_cr_line_month, dtype: object

In [134]:
df['earliest_cr_line_year'].unique()

array(['99', '96', '04', '05', '07', '98', '89', '03', '91', '93', '01',
       '97', '83', '85', '02', '84', '06', '81', '00', '94', '95', '92',
       '86', '90', '88', '80', '08', '73', '79', '87', '82', '78', '71',
       '72', '70', '69', '75', '76', '77', '68', '74', '67', '65', '63',
       '66', '62', '56', '46', '50', '64', '54'], dtype=object)

In [135]:
ts = ['00','01','02','03','04','05','06','07','08','09','10','11','12','13','14','15','16','17','18','19','20','21','22','23','24']

In [136]:
df['earliest_cr_line_year'] = df['earliest_cr_line_year'].apply(lambda x: '20'+ x if x in ts else '19' + x)

In [137]:
df['earliest_cr_line_year'].unique()

array(['1999', '1996', '2004', '2005', '2007', '1998', '1989', '2003',
       '1991', '1993', '2001', '1997', '1983', '1985', '2002', '1984',
       '2006', '1981', '2000', '1994', '1995', '1992', '1986', '1990',
       '1988', '1980', '2008', '1973', '1979', '1987', '1982', '1978',
       '1971', '1972', '1970', '1969', '1975', '1976', '1977', '1968',
       '1974', '1967', '1965', '1963', '1966', '1962', '1956', '1946',
       '1950', '1964', '1954'], dtype=object)

In [115]:
df['revol_util'].head()

1     9.40%
3       21%
4    53.90%
5    28.30%
6    85.60%
Name: revol_util, dtype: object

### <span style="color:green"> Let's remove %age from this column

In [117]:
df['revol_util'] = df['revol_util'].apply(lambda x: float(x.rstrip('%')))

In [118]:
df['revol_util'].head()

1     9.4
3    21.0
4    53.9
5    28.3
6    85.6
Name: revol_util, dtype: float64

In [120]:
df['initial_list_status'].unique()

array(['f'], dtype=object)

### <span style="color:green"> This column has a single value in it and thus, not useful for analysis. We can drop the column.

In [121]:
df = df.drop('initial_list_status', axis=1)

In [122]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 37152 entries, 1 to 39716
Data columns (total 51 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       37152 non-null  int64  
 1   member_id                37152 non-null  int64  
 2   loan_amnt                37152 non-null  int64  
 3   funded_amnt              37152 non-null  int64  
 4   funded_amnt_inv          37152 non-null  float64
 5   term                     37152 non-null  int64  
 6   int_rate                 37152 non-null  float64
 7   installment              37152 non-null  float64
 8   grade                    37152 non-null  object 
 9   sub_grade                37152 non-null  object 
 10  emp_title                37152 non-null  object 
 11  emp_length               37152 non-null  object 
 12  home_ownership           37152 non-null  object 
 13  annual_inc               37152 non-null  float64
 14  verification_status      37

In [123]:
df['last_pymnt_d'].head()

1    Apr-13
3    Jan-15
4    May-16
5    Jan-15
6    May-16
Name: last_pymnt_d, dtype: object

### <span style="color:green"> We can separate month/year for this column. Also, we will rewrite the year for readability.

In [129]:
df['last_pymnt_d_month'] = df['last_pymnt_d'].apply(lambda x: x.split('-')[0])
df['last_pymnt_d_year'] = df['last_pymnt_d'].apply(lambda x: x.split('-')[1])

In [138]:
df['last_pymnt_d_month'].head()

1    Apr
3    Jan
4    May
5    Jan
6    May
Name: last_pymnt_d_month, dtype: object

In [139]:
df['last_pymnt_d_year'].unique()

array(['13', '15', '16', '12', '14', '11', '10', '09', '08'], dtype=object)

In [140]:
df['last_pymnt_d_year'] = df['last_pymnt_d_year'].apply(lambda x: '20'+ x)

In [141]:
df['last_pymnt_d_year'].unique()

array(['2013', '2015', '2016', '2012', '2014', '2011', '2010', '2009',
       '2008'], dtype=object)

In [142]:
df['last_credit_pull_d']

1        Sep-13
3        Apr-16
4        May-16
5        Jan-16
6        May-16
          ...  
39710    Jun-07
39711    Nov-14
39712    Jun-10
39713    Jul-10
39716    Jun-10
Name: last_credit_pull_d, Length: 37152, dtype: object

### <span style="color:green"> We can separate month/year for this column. Also, we will rewrite the year for readability.

In [143]:
df['last_credit_pull_d_month'] = df['last_credit_pull_d'].apply(lambda x: x.split('-')[0])
df['last_credit_pull_d_year'] = df['last_credit_pull_d'].apply(lambda x: x.split('-')[1])

In [144]:
df['last_credit_pull_d_month'].head()

1    Sep
3    Apr
4    May
5    Jan
6    May
Name: last_credit_pull_d_month, dtype: object

In [145]:
df['last_credit_pull_d_year'].unique()

array(['13', '16', '14', '15', '12', '11', '10', '09', '08', '07'],
      dtype=object)

In [146]:
df['last_credit_pull_d_year'] = df['last_credit_pull_d_year'].apply(lambda x: '20'+ x)

In [147]:
df['last_credit_pull_d_year'].unique()

array(['2013', '2016', '2014', '2015', '2012', '2011', '2010', '2009',
       '2008', '2007'], dtype=object)

In [148]:
df['application_type'].head()

1    INDIVIDUAL
3    INDIVIDUAL
4    INDIVIDUAL
5    INDIVIDUAL
6    INDIVIDUAL
Name: application_type, dtype: object

In [149]:
df['application_type'].unique()

array(['INDIVIDUAL'], dtype=object)

### <span style="color:green"> This column has a single value in it and thus can be dropped.

In [150]:
df = df.drop('application_type',axis=1)

In [151]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 37152 entries, 1 to 39716
Data columns (total 54 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   id                        37152 non-null  int64  
 1   member_id                 37152 non-null  int64  
 2   loan_amnt                 37152 non-null  int64  
 3   funded_amnt               37152 non-null  int64  
 4   funded_amnt_inv           37152 non-null  float64
 5   term                      37152 non-null  int64  
 6   int_rate                  37152 non-null  float64
 7   installment               37152 non-null  float64
 8   grade                     37152 non-null  object 
 9   sub_grade                 37152 non-null  object 
 10  emp_title                 37152 non-null  object 
 11  emp_length                37152 non-null  object 
 12  home_ownership            37152 non-null  object 
 13  annual_inc                37152 non-null  float64
 14  verificatio

### <span style="color:red"> We have taken care of Null/Missing values

### <span style="color:red"> Also, we have taken care of data types & splitted few columns to make new ones which will be beneficial for the analysis

## <span style="color:blue"> Next, we have to analyze the features

### <span style="color:green"> Univariate Analysis

In [154]:
df.describe()

id     member_id     loan_amnt   funded_amnt  \
count  3.715200e+04  3.715200e+04  37152.000000  37152.000000   
mean   6.824772e+05  8.495549e+05  11282.064223  11009.697325   
std    2.107507e+05  2.659951e+05   7418.376794   7153.710011   
min    5.574200e+04  7.069900e+04    500.000000    500.000000   
25%    5.161545e+05  6.666318e+05   5600.000000   5500.000000   
50%    6.643465e+05  8.491980e+05  10000.000000  10000.000000   
75%    8.372630e+05  1.047008e+06  15000.000000  15000.000000   
max    1.077430e+06  1.314167e+06  35000.000000  35000.000000   

       funded_amnt_inv          term      int_rate   installment  \
count     37152.000000  37152.000000  37152.000000  37152.000000   
mean      10451.558383     42.505168     12.048491    326.070027   
std        7098.130897     10.668172      3.715893    207.433510   
min           0.000000     36.000000      5.420000     16.080000   
25%        5000.000000     36.000000      9.320000    168.670000   
50%        9000.000000     36.000000     11.860000    283.280000   
75%       14400.000000     60.000000     14.590000    432.122500   
max       35000.000000     60.000000     24.400000   1305.190000   

         annual_inc           dti  ...  total_rec_prncp  total_rec_int  \
count  3.715200e+04  37152.000000  ...     37152.000000   37152.000000   
mean   6.926637e+04     13.398852  ...      9886.635191    2289.116639   
std    6.419245e+04      6.653181  ...      7035.561347    2612.539243   
min    4.000000e+03      0.000000  ...         0.000000       6.220000   
25%    4.160000e+04      8.310000  ...      4800.000000     676.950000   
50%    6.000000e+04     13.500000  ...      8026.770000    1375.905000   
75%    8.300000e+04     18.650000  ...     13908.970000    2866.592500   
max    6.000000e+06     29.990000  ...     35000.020000   23563.680000   

       total_rec_late_fee    recoveries  collection_recovery_fee  \
count        37152.000000  37152.000000             37152.000000   
mean             1.339479     92.731692                11.760063   
std              7.233942    680.518659               140.925102   
min              0.000000      0.000000                 0.000000   
25%              0.000000      0.000000                 0.000000   
50%              0.000000      0.000000                 0.000000   
75%              0.000000      0.000000                 0.000000   
max            180.200000  29623.350000              7002.190000   

       last_pymnt_amnt  policy_code  acc_now_delinq  delinq_amnt  \
count     37152.000000      37152.0         37152.0      37152.0   
mean       2720.281175          1.0             0.0          0.0   
std        4454.529812          0.0             0.0          0.0   
min           0.000000          1.0             0.0          0.0   
25%         224.740000          1.0             0.0          0.0   
50%         564.360000          1.0             0.0          0.0   
75%        3412.667500          1.0             0.0          0.0   
max       36115.200000          1.0             0.0          0.0   

       pub_rec_bankruptcies  
count          37152.000000  
mean               0.040752  
std                0.198667  
min                0.000000  
25%                0.000000  
50%                0.000000  
75%                0.000000  
max                2.000000  

[8 rows x 31 columns]